common/src/gas_provider/node.rs

```diff
@@ -18,6 +18,32 @@
 
 use super::*;
 use codec::MaxEncodedLen;
+use gear_core::ids::ReservationId;
+
+/// ID of the [`GasNode`].
+#[derive(
+    Debug, Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Encode, Decode, TypeInfo, derive_more::From,
+)]
+pub enum GasNodeId {
+    Message(MessageId),
+    Reservation(ReservationId),
+}
+
+impl GasNodeId {
+    pub fn as_message_id(self) -> Option<MessageId> {
+        match self {
+            GasNodeId::Message(message_id) => Some(message_id),
+            GasNodeId::Reservation(_) => None,
+        }
+    }
+
+    pub fn as_reservation_id(self) -> Option<ReservationId> {
+        match self {
+            GasNodeId::Message(_) => None,
+            GasNodeId::Reservation(reservation_id) => Some(reservation_id),
+        }
+    }
+}
```

Comment:
I'd provide `TryInto` instead of this fns
It's also against rust naming concepts to use `fn as_ ... ` with `(self)` arg

Reply:
> I'd provide TryInto instead of this fns

`TryInto` is about `Result` which is inconvenient, so I'll rename methods